### PR TITLE
451006 dmitry ignatchik

### PR DESCRIPTION
**Pull request by MrBlacky01**
Лаба тип сделана :D

### CODE REVIEW

File: TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                        + tempMethod.Info.ParamsNumber.ToString()+"\n");
+
+                }
+            }
+            Console.WriteLine(builder.ToString());
+        }
+
+        private string GetSpaces(int count)
+        {
+            StringBuilder builder = new StringBuilder();
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: @MrBlacky01 You could try to use `string` constructor instead.

File: TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        private readonly static object locker = new object();
+
+        public void Format(TraceResult traceResult)
+        {
+            StringBuilder builder = new StringBuilder();
+            builder.Append("root\n");
+            int countOfThreads;
+
+            lock(locker)
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: @MrBlacky01 Mmm, why are you using locks here?

File: TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            {
+                countOfThreads = traceResult.threadList.Values.Count();
+            }
+
+            ThreadNode tempTreadNode;
+            
+            for(int i = 0; i < countOfThreads; i++)
+            {
+                lock(locker)
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: @MrBlacky01 And here.

File: TracerProgram/Tracer/Formatters/XMLTraceResultFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            fileName = path;
+        }
+
+        public void Format(TraceResult traceResult)
+        {
+            var document = new XDocument();
+            var root = new XElement("root");
+            List<ThreadNode> treadNodeList;
+
+            lock (locker)
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/Formatters/XMLTraceResultFormatter.cs]: @MrBlacky01 Locks?

File: TracerProgram/Tracer/MethodNode.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tracer
+{
+    internal class MethodNode
+    {
+        public MethodInfo Info;
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/MethodNode.cs]: @MrBlacky01 Replace with methods.

File: TracerProgram/Tracer/TreadNode.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        public void AddMethod(MethodInfo method)
+        {
+            if (MethodsTree != null)
+            {
+                ValueOfWorkingMethods++;
+                MethodsTree.Add(new MethodNode(method,ValueOfWorkingMethods));
+            }
+            else
+            {
+                throw new NullReferenceException("List of Methods is empty");
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/TreadNode.cs]: @MrBlacky01 This exception is semantically incorrect as it represents the bug in the program. Replace it with a more appropriate one, for ex. `InvalidOperationException`

File: TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs (hunk 2 above)
Comment (MrBlacky01) [inline TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: I think, that if user want to display result, but tracer still work, don't sitation arises , when 2 treads use traceResult) OK, I was wrong , i all rewrite with locks

File: TracerProgram/Tracer/MethodNode.cs (hunk 5 above)
Comment (MrBlacky01) [inline TracerProgram/Tracer/MethodNode.cs]: What it's mean ? MethodNode is class that contains information about method, and work information about method ?

Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/MethodNode.cs]: @MrBlacky01 Oops. Sorry. User `properties` instead of fields.

**Comment by MrBlacky01** [comment]
Add changes ...

File: TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs [hunk 7]
```diff
@@ -45,10 +37,10 @@ public void Format(TraceResult traceResult)
 
         private string GetSpaces(int count)
         {
-            StringBuilder builder = new StringBuilder();
+            String builder = String.Empty;
```
Comment (RamanBut-Husaim) [inline TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: @MrBlacky01 This is a definitely incorrect fix. String has a constructor that accept a symbol and a number of its repetitions. Be ready to answer questions where to use `StringBuilder` vs `string`.

Comment (MrBlacky01) [inline TracerProgram/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: Ok, I'm fool , i understand you, sorry) few minutes to correct

**Comment by MrBlacky01** [comment]
I'm correct this mistake

**Pull request merged**